urdf/src/model.cpp

```diff
@@ -124,59 +146,55 @@ bool Model::initXml(TiXmlElement * robot_xml)
   return Model::initString(ss.str());
 }
 
-bool Model::initString(const std::string & xml_string)
+bool Model::initString(const std::string & data)
 {
   urdf::ModelInterfaceSharedPtr model;
 
-  // necessary for COLLADA compatibility
-  if (IsColladaData(xml_string)) {
-    fprintf(stderr, "Parsing robot collada xml string is not yet supported.\n");
-    return false;
-    /*
-    ROS_DEBUG("Parsing robot collada xml string");
-
-    static boost::mutex PARSER_PLUGIN_LOCK;
-    static boost::scoped_ptr<pluginlib::ClassLoader<urdf::URDFParser> > PARSER_PLUGIN_LOADER;
-    boost::mutex::scoped_lock _(PARSER_PLUGIN_LOCK);
-
-    try
-    {
-      if (!PARSER_PLUGIN_LOADER)
-        PARSER_PLUGIN_LOADER.reset(new pluginlib::ClassLoader<urdf::URDFParser>("urdf_parser_plugin", "urdf::URDFParser"));
-     const std::vector<std::string> &classes = PARSER_PLUGIN_LOADER->getDeclaredClasses();
-     bool found = false;
-     for (std::size_t i = 0 ; i < classes.size() ; ++i)
-       if (classes[i].find("urdf/ColladaURDFParser") != std::string::npos)
-     {
-       boost::shared_ptr<urdf::URDFParser> instance = PARSER_PLUGIN_LOADER->createInstance(classes[i]);
-       if (instance)
-         model = instance->parse(xml_string);
-       found = true;
-       break;
-     }
-     if (!found)
-       ROS_ERROR_STREAM("No URDF parser plugin found for Collada files. Did you install the corresponding package?");
-   }
-   catch(pluginlib::PluginlibException& ex)
-   {
-     ROS_ERROR_STREAM("Exception while creating planning plugin loader " << ex.what() << ". Will not parse Collada file.");
-   }
+  size_t best_score = std::numeric_limits<size_t>::max();
+  pluginlib::UniquePtr<urdf::URDFParser> best_plugin;
+  std::string best_plugin_name;
+
+  // Figure out what plugins might handle this format
+  for (const std::string & plugin_name : impl_->loader_.getDeclaredClasses()) {
+    pluginlib::UniquePtr<urdf::URDFParser> plugin_instance;
+    try {
+      plugin_instance = impl_->loader_.createUniqueInstance(plugin_name);
+    } catch (const pluginlib::CreateClassException &) {
+      fprintf(stderr, "Failed to load urdf_parser_plugin [%s]\n", plugin_name.c_str());
+      continue;
+    }
+    if (!plugin_instance) {
+      // Debug mode
+      assert(plugin_instance);
+      // Release mode
+      continue;
+    }
+    size_t score = plugin_instance->might_handle(data);
```

Comment:
Should this line ever execute if the `plugin_instance` is `nullptr`?

Reply:
It won't because if `plugin_instance` is `nullptr` the `continue` statement would have skipped to the next loop iteration.

Reply:
:facepalm: right! My brain was reading the `continue` as being confined to the scope of the `if` statement...